puzzlehq/serengeti
# Add grant disbursement to stake to the Puzzle Validator

## Overview

The goal of this new program is to allow us to stake our grant funds natively to our validator.

## Flows
#### Initialization
Like the other grant disbursement programs, initialization of the grant program involves calling the `initialize` function and then funding the grant by calling `process_grant`, ensuring the caller has at least `CREDITS_AMOUNT` to fund the program with.

#### Rewards
Withdrawing rewards involves calling the `withdraw_rewards` function with the amount of microcredits to claim. The rewards will be unbonded from the validator (assuming the remaining total stake does not go below the principal amount).  To claim the rewards from the program, we must appropriately call `claim_unbond_public` and then `withdraw`.

#### Principal
Basically the same as withdrawing rewards, except `withdraw_principal` calculates the valid maximum withdrawal according to the vesting schedule, which involves `time_oracle.aleo`'s current timestamp and the grant's `cliff_timestamp` and `fully_vested_timestamp`.

## Testing

Take a look at the below flow.
https://testnet.aleoscan.io/program?id=grant_disbursement_puzzle_v9.aleo

For testing the `withdraw_principal` flow:

1. bump the program version
2. Add env vars (ping if you need private keys)
3. `leo build`
4. `leo deploy` 

Note: before deploying, make sure you have the keys for the GRANTEE_REWARDS_ADDRESS and the GRANTEE_PRINCIPAL_ADDRESS that are set

5. `leo execute initialize --program grant_disbursement_puzzle_v{your_version}.aleo --broadcast`
6. `leo execute process_grant 0u8 10000000000u64 --program grant_disbursement_puzzle_{your_version}.aleo --broadcast` 

Note: you need 10k credits in the account for step 6

7. `leo execute withdraw_principal 0u8 10000000000u64 --program grant_disbursement_puzzle_{your_version}.aleo --broadcast`
8. wait 360 blocks
9. `leo execute claim_withdrawal_public 10000000000u64 --program grant_disbursement_puzzle_{your_version}.aleo --broadcast`
10. You should have your 10k credits back in the `GRANTEE_PRINCIPAL_ADDRESS`

For testing the `withdraw_rewards` flow:
1. Adjust the CLIFF_TIMESTAMP to > 3000000000u64
2. bump the program version
3. Add env vars (ping if you need private keys)
4. `leo build`
5. `leo deploy` 

Note: before deploying, make sure you have the keys for the GRANTEE_REWARDS_ADDRESS and the GRANTEE_PRINCIPAL_ADDRESS that are set

6. `leo execute initialize --program grant_disbursement_puzzle_v{your_version}.aleo --broadcast`
7. `leo execute process_grant 0u8 10000000000u64 --program grant_disbursement_puzzle_{your_version}.aleo --broadcast` 

Note: you need 10k credits in the account for step 7

8. `leo execute withdraw_rewards 0u8 1u64 --program grant_disbursement_puzzle_{your_version}.aleo --broadcast`
9. wait 360 blocks
10. `leo execute claim_withdrawal_public 1u64 --program grant_disbursement_puzzle_{your_version}.aleo --broadcast`
11. You should have received the 1 microcredit back in the `GRANTEE_PRINCIPAL_ADDRESS`
